internal/discovery/discovery.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2023 The Kubernetes Authors All rights reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"k8s.io/kube-state-metrics/v2/pkg/options"
+)
+
+// NewDiscoveryClientFromOptions creates a new discovery client from the given options.
+func NewDiscoveryClientFromOptions(options *options.Options) *discovery.DiscoveryClient {
+	config, err := clientcmd.BuildConfigFromFlags(options.Apiserver, options.Kubeconfig)
+	if err != nil {
+		panic(err)
+	}
+	discoveryClient := discovery.NewDiscoveryClientForConfigOrDie(config)
+	return discoveryClient
+}
+
+// fetchObjects fetches all the objects that can be listed from the apiserver, every 30 seconds.
+func (r *GVKMaps) fetchObjects(discoveryClient *discovery.DiscoveryClient) {
+	if !r.Heartbeat.IsZero() && time.Since(r.Heartbeat) < 30*time.Second {
+		return
+	}
+	groupList, err := r.Client.ServerGroups()
+	if err != nil {
+		panic(err)
+	}
+	gToVMap := make(map[string][]string)
+	gToKMap := make(map[string][]string)
+	gVKMap := make(map[[2]string][]string)
+	gKVMap := make(map[[2]string][]string)
```

Comment:
Do we really want to optimize lookup that much? It sounds a bit overkill and would most likely mean that we have to store the data multiple times.

A simpler approach could be:
```
map[G]map[V]K
```

That would allow for easy lookup, and the worst scenario is when a GK is specified. In that case we would just have to go through all versions to find the ones where the kind exist. This isn't too bad as in practice there shouldn't be many versions for a group.